File: src/builtins.js
var {BuiltinError} = require('./error');
var fromNow = require('./from-now');
var {
  isString, isNumber, isBool,
  isArray, isObject, isJSON,
  isNull, isFunction,
} = require('./type-utils');

let types = {
  string: isString,
  number: isNumber,
  boolean: isBool,
  array: isArray,
  object: isObject,
  json: isJSON,
  null: isNull,
  function: isFunction,
};

let builtinError = (builtin) => new BuiltinError(`invalid arguments to ${builtin}`);

module.exports = (context) => {
  let builtins = {};

  let define = (name, context, {
    argumentTests = [],
    minArgs = false,
    variadic = null,
    invoke,
  }) => context[name] = (...args) => {
    if (!variadic && args.length < argumentTests.length) {
      throw builtinError(`builtin: ${name}`, `${args.toString()}, too few arguments`);
    }

    if (minArgs && args.length < minArgs) {
      throw builtinError(`builtin: ${name}: expected at least ${minArgs} arguments`);
    }

    if (variadic) {
      argumentTests = args.map(() => variadic);
    }

    args.forEach((arg, i) => {
      if (!argumentTests[i].split('|').some(test => types[test](arg))) {
        throw builtinError(`builtin: ${name}`, `argument ${i + 1} to be ${argumentTests[i]} found ${typeof arg}`);
      }
    });

    return invoke(...args);
  };

  // Math functions
  ['max', 'min'].forEach(name => {
    if (Math[name] == undefined) {
      throw new Error(`${name} in Math undefined`);
    }
    define(name, builtins, {
      minArgs: 1,
      variadic: 'number',
      invoke: (...args) => Math[name](...args),
    });
  });

  ['sqrt', 'ceil', 'floor', 'abs'].forEach(name => {
    if (Math[name] == undefined) {
      throw new Error(`${name} in Math undefined`);
    }
    define(name, builtins, {
      argumentTests: ['number'],
      invoke: num => Math[name](num),
    });
  });

  // String manipulation
  define('lowercase', builtins, {
    argumentTests: ['string'],
    invoke: str => str.toLowerCase(),
  });

  define('uppercase', builtins, {
    argumentTests: ['string'],
    invoke: str => str.toUpperCase(),
  });

  define('str', builtins, {
    argumentTests: ['string|number|boolean|null'],
    invoke: obj => {
      if (obj === null) {
        return 'null';
      }
      return obj.toString();
    },
  });

  define('len', builtins, {
    argumentTests: ['string|array'],
    invoke: obj => Array.from(obj).length,
  });

  define('strip', builtins, {
    argumentTests: ['string'],
    invoke: str => str.trim(),
  });

  define('rstrip', builtins, {
    argumentTests: ['string'],
    invoke: str => str.replace(/\s+$/, ''),
  });

  define('lstrip', builtins, {
    argumentTests: ['string'],
    invoke: str => str.replace(/^\s+/, ''),
  });

  // Miscellaneous
  define('fromNow', builtins, {
    variadic: 'string',
    minArgs: 1,
    invoke: (str, reference) => fromNow(str, reference || context.now),
  });

  define('typeof', builtins, {
    argumentTests: ['string|number|boolean|array|object|null|function'],
    invoke: x => {
      for (type of ['string', 'number', 'boolean', 'array', 'object', 'function']) {
        if (types[type](x)) {
          return type;
        }
      }
      if (types['null'](x)) {
        return null;
      }
      throw builtinError('builtin: typeof', `argument ${x} to be a valid json-e type. found ${typeof arg}`);
    },
  });

  return Object.assign({}, builtins, context);
};

File: jsone/builtins.py
from __future__ import absolute_import, print_function, unicode_literals

import math
from .interpreter import ExpressionError
from . import shared

builtins = {}


def builtin(name, variadic=None, argument_tests=None, minArgs=None):
    def wrap(fn):
        def bad():
            raise ExpressionError('invalid arguments to {}'.format(name))
        if variadic:
            def invoke(args):
                if minArgs:
                    if len(args) < minArgs:
                        bad()
                for arg in args:
                    if not variadic(arg):
                        bad()
                return fn(*args)

        elif argument_tests:
            def invoke(args):
                if len(args) != len(argument_tests):
                    bad()
                for t, arg in zip(argument_tests, args):
                    if not t(arg):
                        bad()
                return fn(*args)

        else:
            def invoke(args):
                return fn(*args)

        builtins[name] = invoke
        return fn
    return wrap


def number(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool)

def string(v):
    return isinstance(v, basestring)

def string_or_array(v):
    return isinstance(v, (basestring, list))

def anything(v):
    return isinstance(v, (basestring, int, float, list))

# ---


builtin('min', variadic=number, minArgs=1)(min)
builtin('max', variadic=number, minArgs=1)(max)
builtin('sqrt', argument_tests=[number])(math.sqrt)
builtin('ceil', argument_tests=[number])(math.ceil)
builtin('floor', argument_tests=[number])(math.floor)
builtin('abs', argument_tests=[number])(abs)

@builtin('lowercase', argument_tests=[string])
def lowercase(v):
    return v.lower()

@builtin('uppercase', argument_tests=[string])
def lowercase(v):
    return v.upper()

builtin('len', argument_tests=[string_or_array])(len)

@builtin('str', argument_tests=[anything])
def to_str(v):
    if isinstance(v, bool):
        return {True: 'true', False: 'false'}[v]
    elif isinstance(v, list):
        return ','.join(to_str(e) for e in v)
    else:
        return str(v)

@builtin('str', argument_tests=[anything])
def to_str(v):
    if isinstance(v, bool):
        return {True: 'true', False: 'false'}[v]
    elif isinstance(v, list):
        return ','.join(to_str(e) for e in v)
    else:
        return str(v)

builtin('fromNow', argument_tests=[string])(shared.fromNow)
